processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -512,26 +516,28 @@ private long getMaxTimeMillis()
     return getFacts().lastKey().getTimestamp();
   }
 
-  private String[] getDimVals(final DimDim dimLookup, final List<String> dimValues)
+  private int[] getDimVals(final DimDim dimLookup, final List<String> dimValues)
   {
-    final String[] retVal = new String[dimValues.size()];
     if (dimValues.size() == 0) {
       // NULL VALUE
-      if (!dimLookup.contains(null)) {
-        dimLookup.add(null);
-      }
+      dimLookup.add("");
       return null;
     }
+
+    if (dimValues.size() == 1) {
+      return new int[]{dimLookup.add(dimValues.get(0))};
+    }
+
+    // todo: make option for this
+    String[] dimArray = dimValues.toArray(new String[dimValues.size()]);
+    Arrays.sort(dimArray);
+
+    final int[] retVal = new int[dimArray.length];
+
     int count = 0;
-    for (String dimValue : dimValues) {
-      String canonicalDimValue = dimLookup.get(dimValue);
-      if (!dimLookup.contains(canonicalDimValue)) {
-        dimLookup.add(dimValue);
-      }
-      retVal[count] = canonicalDimValue;
-      count++;
+    for (String dimValue : dimArray) {
+      retVal[count++] = dimLookup.add(dimValue);
```

Comment:
Now we stores int[] array instead of String[], it's not need to be canonicalized.